README.md
# react-native-circular-timer

Circular Timer component for React Native

<a href="https://twitter.com/intent/follow?screen_name=meharbhutta">
    <img 
        src="https://img.shields.io/twitter/follow/meharbhutta.svg?style=social&logo=twitter"
        alt="follow on Twitter"
    >
</a>

## NPM

- stable release version: ![version](https://img.shields.io/badge/version-1.1.4-blue.svg?cacheSeconds=2592000)
- package downloads: ![downloads](https://img.shields.io/badge/downloads-22%2Fweek-brightgreen.svg?cacheSeconds=2592000)
- [![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)

## Show Case

![](https://raw.githubusercontent.com/meharbhutta/react-native-circular-timer/master/example/react-native-circular-timer.gif)

## Getting Started

### Installation

```bash
npm i react-native-circular-timer --save
```

### Basic Usage

- Initialization of a react-native project

```bash
$ npx react-native init AwesomeProject
```

### Note: [GUIDE](https://facebook.github.io/react-native/docs/getting-started)

- Then, edit `AwesomeProject/App.js`, like this:

```javascript
import { View } from 'react-native';
import CircularTimer from 'react-native-circular-timer';

export default class App extends Component<Props> {
  _restartTimer = () => {
    if (this._timerRef) this._timerRef.restart();
  };

  render() {
    return (
      <View
        style={{
          flex: 1,
          justifyContent: 'center',
          alignItems: 'center',
          backgroundColor: 'white'
        }}
      >
        <CircularTimer
          ref={refs => (this._timerRef = refs)}
          onTimeElapsed={() => {
            console.log('Timer Finished!');
          }}
          showSecond={true}
        />
        <TouchableOpacity
          onPress={this._restartTimer}
          style={{
            marginTop: 20,
            borderRadius: 6,
            backgroundColor: '#000000'
          }}
        >
          <Text
            style={{
              padding: 16,
              fontSize: 18,
              color: 'white',
              fontWeight: 'bold'
            }}
          >
            Restart Timer
          </Text>
        </TouchableOpacity>
      </View>
    );
  }
}
```

### Props

| parameter             | type     | required | description                          | default     |
| :-------------------- | :------- | :------- | :----------------------------------- | :---------- |
| seconds               | number   | yes      | Time in seconds                      |             |
| radius                | number   | yes      | Radius of the circle                 |             |
| showSecond            | boolean  | no       | To show the `sec` keyword with timer | `false`     |
| borderWidth           | number   | no       | The border width of the circle       | `10`        |
| borderColor           | string   | no       | The color of the border              | `"#0E3657"` |
| borderBackgroundColor | string   | no       | The background color of the border   | `"#A8C3BC"` |
| circleColor           | string   | no       | The color of the circle              | `"#FFF"`    |
| style                 | object   | no       | To apply style to the timer          |             |
| textStyle             | object   | no       | To override the text style           | default     |
| secondStyle           | object   | no       | To override the second style         | default     |
| onTimeElapsed         | function | no       | Callback for the timer end           | () => {}    |

### Methods

| method  | description                |
| :------ | :------------------------- |
| restart | To restart the timer again |

### Note: 
Firstly, you need to create reference of the component and then call the restart method using that reference [GUIDE](https://stackoverflow.com/questions/37949981/call-child-method-from-parent).

### To run example

```bash
cd example
npm install
npx react-native run-android (For android)
npx react-native run-ios (For ios)
```

#### In case of any issue follow the [GUIDE](https://facebook.github.io/react-native/docs/getting-started).

### Supported React Native Versions  

This project aims to support any version of React Native.

If you require new features or bug fixes for older versions you can fork this project.

### Credits

The idea for this modules came from MrToph react-native-countdown-circle package.

### Licenses

CircularTimer - MIT © MeharBhutta


CHANGELOG.md
#### 1.1.4-5 (2019-12-07)

##### Documentation Changes

*  stable release updated ([c4d779e0](https://github.com/meharbhutta/react-native-circular-timer/commit/c4d779e084ce9308f40be0b6b77f31e58ac072fb))

#### 1.1.3-4 (2019-12-07)

##### New Features

* **template:**  github template added ([da45c6b4](https://github.com/meharbhutta/react-native-circular-timer/commit/da45c6b4beda51abb8b3415bbaf45394c1259583))

#### 1.1.2-3 (2019-12-05)

##### Documentation Changes

*  stable release updated ([ccb729a6](https://github.com/meharbhutta/react-native-circular-timer/commit/ccb729a68c726c6f0c5278354e47d83605f5540a))

##### Bug Fixes

* **lock.json:**  removed package-lock.json ([fc0c268e](https://github.com/meharbhutta/react-native-circular-timer/commit/fc0c268ef26749a852aef79e9be95264abc75888))

##### Other Changes

* //github.com/meharbhutta/react-native-circular-timer ([63d6bfe0](https://github.com/meharbhutta/react-native-circular-timer/commit/63d6bfe0dbd2ac1538a5627543d42e047733ca97))

#### 1.1.1-2 (2019-12-05)

##### New Features

*  changelog.md added ([8564a6bc](https://github.com/meharbhutta/react-native-circular-timer/commit/8564a6bcc5a28af1eda272f56c21434f64cb210e))



src/index.d.ts
declare module 'react-native-circular-timer' {
  import { ViewStyle, TextStyle } from 'react-native';

  /**
   * Properties of the CircularTimer component.
   */
  interface CircularTimerProps {
    /**
     * The time in seconds is required.
     */
    seconds: number;

    /**
     * The radius of the circle is required.
     */
    radius: number;

    /**
     * The border width of the circle to indicate time elapsed. The default value is 10.
     */
    borderWidth?: number;

    /**
     * The border color of the circle. The default color is "#0E3657".
     */
    borderColor?: string;

    /**
     * The border background color of the circle. The default color is "#A8C3BC".
     */
    borderBackgroundColor?: string;

    /**
     * The color of the circle. The default color is "#FFF".
     */
    circleColor?: string;

    /**
     * Style for the circular timer.
     */
    style?: ViewStyle;

    /**
     * Style to override the default text style.
     */
    textStyle?: TextStyle;

    /**
     * Callback for the timer end.
     */
    onTimeElapsed?(): void;
  }

  class CircularTimer extends React.Component<CircularTimerProps, {}> {
    restart(): void;
  }

  export default CircularTimer;
}


src/CircularTimer.js
'use strict';

import * as React from 'react';
import PropTypes from 'prop-types';
import { Text, View, StyleSheet, ViewPropTypes } from 'react-native';

const ViewPropTypesStyle = ViewPropTypes ? ViewPropTypes.style : View.propTypes.style;

export default class CircularTimer extends React.Component {
  static propTypes = {
    seconds: PropTypes.number.isRequired,
    radius: PropTypes.number.isRequired,
    showSecond: PropTypes.bool,
    borderWidth: PropTypes.number,
    borderColor: PropTypes.string,
    borderBackgroundColor: PropTypes.string,
    circleColor: PropTypes.string,
    style: ViewPropTypesStyle,
    textStyle: Text.propTypes.style,
    secondStyle: Text.propTypes.style,
    onTimeElapsed: PropTypes.func
  };

  static defaultProps = {
    borderColor: '#0E3657',
    borderBackgroundColor: '#A8C3BC',
    circleColor: '#FFF',
    borderWidth: 10,
    radius: 40,
    seconds: 60,
    showSecond: false,
    style: null,
    textStyle: null,
    secondStyle: null,
    onTimeElapsed: () => {}
  };

  constructor(props) {
    super(props);

    this.state = {
      elapsedTime: props.seconds,
      rotate: 0
    };
  }

  componentDidMount() {
    this._start();
  }

  _clearTimer = () => {
    if (this._timer) clearInterval(this._timer);
  };

  _start = () => {
    this._timer = setInterval(() => {
      const { elapsedTime, rotate } = this.state,
        { seconds, onTimeElapsed } = this.props;
      if (elapsedTime > 0)
        this.setState(
          {
            elapsedTime: elapsedTime - 1,
            rotate: rotate + 360 / seconds
          },
          () => {
            if (elapsedTime - 1 === 0) onTimeElapsed();
          }
        );
      else this._clearTimer();
    }, 1000);
  };

  restart = () => {
    this._clearTimer();
    this.setState(
      {
        elapsedTime: this.props.seconds,
        rotate: 0
      },
      () => {
        this._start();
      }
    );
  };

  componentWillUnmount() {
    this._clearTimer();
  }

  render() {
    const { elapsedTime, rotate } = this.state,
      {
        seconds,
        showSecond,
        borderColor,
        borderBackgroundColor,
        circleColor,
        style,
        textStyle,
        secondStyle,
        radius,
        borderWidth
      } = this.props,
      check = elapsedTime > seconds / 2,
      leftHalfBackColor = check ? borderColor : 'transparent',
      rightHalfBackColor = check ? 'transparent' : borderBackgroundColor,
      fullCircle = {
        width: radius * 2,
        height: radius * 2,
        borderRadius: radius
      },
      halfCircle = {
        width: radius,
        height: radius * 2,
        borderRadius: radius
      },
      innerCircle = {
        width: (radius - borderWidth) * 2,
        height: (radius - borderWidth) * 2,
        borderRadius: radius - borderWidth,
        flex: 0
      };

    return (
      <View style={{ ...style, ...__styles.container, ...fullCircle, backgroundColor: borderBackgroundColor }}>
        <View
          style={{
            ...__styles.move,
            transform: [{ rotate: `${rotate}deg` }],
            ...fullCircle
          }}
        >
          <View
            style={{
              ...__styles.halfCircle,
              ...__styles.rightHalfCircle,
              ...halfCircle,
              backgroundColor: borderColor
            }}
          />
        </View>
        <View style={{ ...__styles.move, ...fullCircle }}>
          <View
            style={{
              ...__styles.halfCircle,
              ...__styles.leftHalfCircle,
              ...__styles.rightMove,
              ...halfCircle,
              backgroundColor: leftHalfBackColor
            }}
          />
          <View
            style={{
              ...__styles.halfCircle,
              ...__styles.rightHalfCircle,
              ...__styles.rightMove,
              ...halfCircle,
              backgroundColor: rightHalfBackColor
            }}
          />
        </View>
        <View style={{ ...__styles.circle, ...innerCircle, backgroundColor: circleColor }}>
          <Text style={textStyle}>{elapsedTime}</Text>
          {showSecond && <Text style={secondStyle}>sec</Text>}
        </View>
      </View>
    );
  }
}

const __styles = StyleSheet.create({
  container: {
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'center',
    overflow: 'hidden'
  },
  circle: {
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 8
  },
  move: {
    position: 'absolute',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'transparent'
  },
  halfCircle: {
    position: 'absolute',
    overflow: 'hidden',
    zIndex: 1,
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0
  },
  rightMove: {
    zIndex: 4
  },
  leftHalfCircle: {
    left: 0,
    transform: [{ rotate: '180deg' }]
  },
  rightHalfCircle: {
    right: 0
  }
});


example/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 */

import React, { Component } from 'react';
import { StyleSheet, View, TouchableOpacity, Text } from 'react-native';
import CircularTimer from 'react-native-circular-timer';

type Props = {};
export default class App extends Component<Props> {
  _restartTimer = () => {
    if (this._timerRef) this._timerRef.restart();
  };
  
  render() {
    return (
      <View style={styles.container}>
        <CircularTimer
          ref={refs => (this._timerRef = refs)}
          onTimeElapsed={() => {
            console.log('Timer Finished!');
          }}
          showSecond={true}
        />
        <TouchableOpacity onPress={this._restartTimer} style={styles.button}>
          <Text style={styles.restart}>Restart Timer</Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white'
  },
  button: {
    marginTop: 20,
    borderRadius: 6,
    backgroundColor: '#000000'
  },
  restart: {
    padding: 16,
    fontSize: 18,
    color: 'white',
    fontWeight: 'bold'
  }
});
